### PR TITLE
[Beam-6627] add metrics to file based tests

### PR DESCRIPTION
**Pull request by mwalenia**
Similarly as in my earlier [PR](https://github.com/apache/beam/pull/7772), I added metric reporting to the rest of file based IO tests.

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by mwalenia** [comment]
R: @lgajowy

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/avro/AvroIOIT.java [hunk 1]
```diff
@@ -125,7 +144,40 @@ public void writeThenReadAll() {
         ParDo.of(new DeleteFileFn())
             .withSideInputs(consolidatedHashcode.apply(View.asSingleton())));
 
-    pipeline.run().waitUntilFinish();
+    PipelineResult result = pipeline.run();
+    result.waitUntilFinish();
+    gatherAndPublishMetricResults(result);
+  }
+
+  private void gatherAndPublishMetricResults(PipelineResult result) {
```
Comment (lgajowy) [inline sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/avro/AvroIOIT.java]: Do you think we could reduce the duplication here? Given we have similar code in all tests, maybe we can have it in one place (namely `gatherAndPublishMetricResults` and `extractMetrics`.

**Comment by mwalenia** [comment]
@lgajowy it's ready for review now, WDYT?

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/IOITMetrics.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      Set<Function<MetricsReader, NamedTestResult>> metricSuppliers,
+      PipelineResult result,
+      String namespace) {
+    this.metricSuppliers = metricSuppliers;
+    this.result = result;
+    this.namespace = namespace;
+  }
+
+  public void publish(String bigQueryDataset, String bigQueryTable) {
+    String uuid = UUID.randomUUID().toString();
```
Comment (lgajowy) [inline sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/IOITMetrics.java]: This should be passed to IOITMetrics from the test class (uuid, timestamp), right?

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/avro/AvroIOIT.java [hunk 3]
```diff
@@ -125,7 +144,50 @@ public void writeThenReadAll() {
...
+    String timestamp = Timestamp.now().toString();
+
+    Set<Function<MetricsReader, NamedTestResult>> metricSuppliers =
+        fillMetricSuppliers(uuid, timestamp);
+    new IOITMetrics(metricSuppliers, result, AVRO_NAMESPACE)
+        .publish(bigQueryDataset, bigQueryTable);
+  }
+
+  private Set<Function<MetricsReader, NamedTestResult>> fillMetricSuppliers(
+      String uuid, String timestamp) {
```
Comment (lgajowy) [inline sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/avro/AvroIOIT.java]: This uuid and timestamp will be different than the ones from IOITMetrics. They should be the same

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Java TextIO Performance Test HDFS

**Comment by mwalenia** [comment]
Run Java CompressedTextIO Performance Test HDFS

**Comment by mwalenia** [comment]
Run Java ManyFilesTextIO Performance Test HDFS

**Comment by mwalenia** [comment]
Run Java TextIO Performance Test

**Comment by mwalenia** [comment]
Run Java CompressedTextIO Performance Test

**Comment by mwalenia** [comment]
Run Java ManyFilesTextIO Performance Test

**Comment by mwalenia** [comment]
Run Java AvroIO Performance Test

**Comment by mwalenia** [comment]
Run Java XmlIO Performance Test

**Comment by mwalenia** [comment]
Run Java ParquetIO Performance Test

**Comment by mwalenia** [comment]
Run Java ParquetIO Performance Test HDFS

**Comment by mwalenia** [comment]
Run Java XmlIO Performance Test HDFS

**Comment by mwalenia** [comment]
Run Java AvroIO Performance Test HDFS

**Comment by mwalenia** [comment]
@lgajowy  All the jobs have been triggered and changes were applied, WDYT?

**Comment by lgajowy** [comment]
@mwalenia I took a look into bigquery but was unable to find the tables that you specified. That probably means that nothing gets published. Could you examine this further? This looks like some oversight during refactoring 🤔

**Comment by lgajowy** [comment]
Alternatively, the seed job was reloaded from master before you started AvroIOIT. Let's check:

**Comment by lgajowy** [comment]
Run seed job

**Comment by lgajowy** [comment]
Run Java AvroIO Performance Test

**Comment by lgajowy** [comment]
Run Java ParquetIO Performance Test

**Comment by lgajowy** [comment]
Run Java XmlIO Performance Test

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/IOITMetrics.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.Collection;
+import java.util.Set;
+import java.util.function.Function;
+import org.apache.beam.sdk.PipelineResult;
+import org.apache.beam.sdk.testutils.NamedTestResult;
+import org.apache.beam.sdk.testutils.metrics.MetricsReader;
+import org.apache.beam.sdk.testutils.publishing.BigQueryResultsPublisher;
+import org.apache.beam.sdk.testutils.publishing.ConsoleResultPublisher;
+
+public class IOITMetrics {
```
Comment (lgajowy) [inline sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/IOITMetrics.java]: We're missing a Javadoc here. This is why java precommit fails (chechstyle error). Could you add the docs here describing what does the class do?

**Comment by lgajowy** [comment]
It looks like it was the problem I described - periodically-fired seed_job "reseeded" jobs before the avro, parquet and xml tests were fired... 🙉  It won't happen after the changes are merged.

Please fix the checkstyle problem, rebuild branch history and I think we'll be good to go. :)

**Comment by mwalenia** [comment]
Run Java PreCommit

**Comment by lgajowy** [comment]
LGTM, merging

**Comment by lgajowy** [comment]
@mwalenia please remember to add dashboards for those tests. Same as we have for TextIOIT: https://apache-beam-testing.appspot.com/explore?dashboard=5629522644828160

**Pull request merged**